modules/ega_download/1.0/ega_download.smk

```diff
@@ -0,0 +1,176 @@
+    params:
+        credentials = CFG["credentials_file"],
+        additional_args = CFG["options"]["additional_args"]
+    conda:
+        CFG["conda_envs"]["pyega3"]
+    threads:
+        CFG["threads"]["ega_file_download"]
+    resources:
+        **CFG["resources"]["ega_file_download"]
+    wildcard_constraints:
+        file_format = "|".join(CFG["samples"]["file_format"].tolist())
+    shell:
+        op.as_one_line("""
+        pyega3
+        {params.additional_args}
+        -cf {params.credentials}
+        fetch {wildcards.egaf}
+        --output-dir $(dirname "$(dirname {output.ega_file})")
+        --max-retries -1
+        --retry-wait 30
+        > {log.stdout}
+        2> {log.stderr}
+        &&
+        mv $(dirname {output.ega_file})/{wildcards.file_name}.{wildcards.file_format} {output.ega_file}
+        """)
+
+# function to get ega_file files that drops EGA IDs from the name of the final
+# file and uses sample_id instead
+def get_ega_file (wildcards):
+    CFG = config["lcr-modules"]["ega_download"]
+    tbl = CFG["samples"]
+    this_sample = tbl[(tbl.seq_type == wildcards.seq_type) & (tbl.sample_id == wildcards.sample_id) & (tbl.file_format == wildcards.file_format)]
+    if (this_sample.shape[0]>0):
+        this_sample = this_sample.iloc[int(wildcards.read) - 1]
+    this_egas, this_egad, this_egan, this_egaf, this_name = this_sample['EGAS'], this_sample['EGAD'], this_sample['EGAN'], this_sample['EGAF'], this_sample['file_name']
+    this_file = expand(
+            str(
+                CFG["dirs"]["ega_download"] + "{{seq_type}}/{{file_format}}/{{study_id}}/{egaf}/{egas}--{egad}--{egan}--{egaf}--{{file_name}}.{{file_format}}"
+            ),
+            egas = this_egas,
+            egad = this_egad,
+            egan = this_egan,
+            egaf = this_egaf
+        )[0]
+    this_file = str(this_file).replace("{file_name}", this_name)
+    return(this_file)
+
+# Symlinks the final output files into the module results directory (under '99-outputs/')
+rule _ega_output_files:
+    input:
+        ega_file = get_ega_file
+    output:
+        ega_file = CFG["dirs"]["outputs"] + "{seq_type}/{study_id}/{sample_id}{read}.{file_format}"
+    wildcard_constraints:
+        file_format = "|".join(CFG["samples"]["file_format"].tolist())
+    run:
+        op.relative_symlink(input.ega_file, output.ega_file, in_module=True)
+
+# Generates the target files for each run, which generate the symlinks
+rule _ega_all:
+    input:
+        expand(
+            [
+                str(rules._ega_output_files.output.ega_file)
+            ],
+            zip,  # Run expand() with zip(), not product()
+            sample_id=CFG["samples"]["sample_id"],
+            file_format=CFG["samples"]["file_format"],
+            seq_type=CFG["samples"]["seq_type"],
+            study_id=[CFG["study_id"]]*len(CFG["samples"]["file_name"]),
+            read = [""]
+        ) if out_file_type in ["cram", "bam"] else expand(
```

Comment:
I've never seen an ifelse formatted like this :o  Does it not work if the `if` is stated at the top, instead of after the first `expand()` as it is here? 
Nothing wrong with this, I'm just curious about it

Reply:
haha, I copied this from one of the other modules some time ago and follow it since then - I think this is a more "pythonic" way of a one-liner. The standard way should work too, I think :smile:

Reply:
great to know!